design/clean_artifacts_in_csi_flow.md

```diff
@@ -0,0 +1,422 @@
+
+As a result, the VolumeSnapshotClass should still be synced in the backup sync process.
+
+### Backup Deletion
+Two factors are worthy for consideration for the backup deletion change:
+* Because the VolumeSnapshotContents generated by the backup are not synced anymore, and the VolumeSnapshotContents generated during the backup will not be kept too. The backup deletion needs to generate a VolumeSnapshotContent, then deletes it to make sure the snapshots in the storage provider are clean too.
+* The VolumeSnapshots generated by the backup are already deleted in the backup process, we don't need a DeleteItemAction for the VolumeSnapshot anymore. As a result, the `velero.io/csi-volumesnapshot-delete` plugin is unneeded.
+
+For the VolumeSnapshotContent DeleteItemAction, we need to generate a VolumeSnapshotContent. Because we only care about the snapshot deletion, we don't need to create a VolumeSnapshot associated with the VolumeSnapshotContent.
+
+Create a static VolumeSnapshotContent, then point it to a pseudo VolumeSnapshot, and reference to the snapshot handle should be enough.
+
+The following is an example of the implementation.
+``` go
+	snapCont.Spec.DeletionPolicy = snapshotv1api.VolumeSnapshotContentDelete
+
+	snapCont.Spec.Source = snapshotv1api.VolumeSnapshotContentSource{
+		SnapshotHandle: snapCont.Status.SnapshotHandle,
+	}
+
+	snapCont.Spec.VolumeSnapshotRef = corev1api.ObjectReference{
+		APIVersion: snapshotv1api.SchemeGroupVersion.String(),
+		Kind:       "VolumeSnapshot",
+		Namespace:  "ns-" + string(snapCont.UID),
+		Name:       "name-" + string(snapCont.UID),
+	}
+
+	snapCont.ResourceVersion = ""
+
+	if err := p.crClient.Create(context.TODO(), &snapCont); err != nil {
+		return errors.Wrapf(err, "fail to create VolumeSnapshotContent %s", snapCont.Name)
+	}
+
+	// Read resource timeout from backup annotation, if not set, use default value.
+	timeout, err := time.ParseDuration(
+		input.Backup.Annotations[velerov1api.ResourceTimeoutAnnotation])
+	if err != nil {
+		p.log.Warnf("fail to parse resource timeout annotation %s: %s",
+			input.Backup.Annotations[velerov1api.ResourceTimeoutAnnotation], err.Error())
+		timeout = 10 * time.Minute
+	}
+	p.log.Debugf("resource timeout is set to %s", timeout.String())
+
+	interval := 5 * time.Second
+
+	// Wait until VSC created and ReadyToUse is true.
+	if err := wait.PollUntilContextTimeout(
+		context.Background(),
+		interval,
+		timeout,
+		true,
+		func(ctx context.Context) (bool, error) {
+			tmpVSC := new(snapshotv1api.VolumeSnapshotContent)
+			if err := p.crClient.Get(ctx, crclient.ObjectKeyFromObject(&snapCont), tmpVSC); err != nil {
+				return false, errors.Wrapf(
+					err, "failed to get VolumeSnapshotContent %s", snapCont.Name,
+				)
+			}
+
+			if tmpVSC.Status != nil && boolptr.IsSetToTrue(tmpVSC.Status.ReadyToUse) {
```

Comment:
Since we are creating VSC from an existing snapshot handle, we can directly set it as ReadyToUse

Reply:
Because the `ReadyToUse` is in the VSC's status section, we cannot set its value in creation.
It seems we still need to wait for the state change.